src/lpython/semantics/python_ast_to_asr.cpp

```diff
@@ -3740,7 +3752,44 @@ class CommonVisitor : public AST::BaseVisitor<Struct> {
             }
             return final_result;
         } else {
+            ASR::expr_t *index = nullptr;
             this->visit_expr(*m_slice);
+            if (ASR::is_a<ASR::Const_t>(*type)) {
+                ASR::ttype_t *contained_type = ASRUtils::type_get_past_const(type);
+                if (ASR::is_a<ASR::Dict_t>(*contained_type)) {
+                    index = ASRUtils::EXPR(tmp);
+                    ASR::ttype_t *key_type = ASR::down_cast<ASR::Dict_t>(contained_type)->m_key_type;
+                    if (!ASRUtils::check_equal_type(ASRUtils::expr_type(index), key_type)) {
+                        throw SemanticError("Key type should be '" + ASRUtils::type_to_str_python(key_type) +
+                                                "' instead of '" +
+                                                ASRUtils::type_to_str_python(ASRUtils::expr_type(index)) + "'",
+                                            index->base.loc);
+                    }
+                    tmp = ASR::make_DictItem_t(al, loc, value, index, nullptr,
+                                    ASR::down_cast<ASR::Dict_t>(contained_type)->m_value_type, nullptr);
+                    return false;
+                }
+                else if (ASR::is_a<ASR::List_t>(*contained_type)) {
+                    this->visit_expr(*m_slice);
+                    index = ASRUtils::EXPR(tmp);
+                    ASR::expr_t* val = ASRUtils::expr_value(index);
+                    if (val && ASR::is_a<ASR::IntegerConstant_t>(*val)) {
+                        if (ASR::down_cast<ASR::IntegerConstant_t>(val)->m_n < 0) {
+                            // Replace `x[-1]` to `x[len(x)+(-1)]`
+                            ASR::ttype_t *int_type = ASRUtils::TYPE(ASR::make_Integer_t(
+                                                            al, loc, 4));
+                            ASR::expr_t *list_len = ASRUtils::EXPR(ASR::make_ListLen_t(
+                                        al, loc, value, int_type, nullptr));
+                            ASR::expr_t *neg_idx = ASRUtils::expr_value(index);
+                            index = ASRUtils::EXPR(ASR::make_IntegerBinOp_t(al, loc,
+                                list_len, ASR::binopType::Add, neg_idx, int_type, nullptr));
+                        }
+                    }
+                    tmp = make_ListItem_t(al, loc, value, index,
+                                      ASR::down_cast<ASR::List_t>(contained_type)->m_type, nullptr);
+                    return false;
+                }
+            }
```

Comment:
```suggestion
            this->visit_expr(*m_slice);
            type = ASRUtils::type_get_past_const(type);
```

Does simply doing the above help?

Reply:
Amazing @Shaikh-Ubaid! It works and I am truly speechless now!

If only I knew about `type_get_past_const()` handling non-constant types, I would have not struggled so much. But I think it was not all in vain. I learnt a lot of things, including this one-liner fix. :smile:

Reply:
The changes in the `asr_to_llvm.cpp` file is required though.

Also, we will have to figure out a way to make slicing work with this method. Currently it fails with this error:
```console
(lp) saurabh-kumar@Awadh:~/Projects/System/lpython$ ./src/bin/lpython ./examples/example.py
LCOMPILERS_ASSERT failed: /home/saurabh-kumar/Projects/System/lpython/src/libasr/../libasr/asr_utils.h
function Call_t_body(), line number 4596 at 
ASRUtils::check_equal_type(arg_type, orig_arg_type)
ERROR MESSAGE:
ASRUtils::check_equal_type(Const[list[i32]], list[i32])
.
.
.
```